### ISSUE TITLE
"Server is not listening on port 3000" but application is running.

### ISSUE THREAD

**Issue by TheBeachMaster** [sveltejs/sapper#730]
This  is **sort of** similar to [this issue](https://github.com/sveltejs/sapper/issues/431) and [this issue](https://github.com/sveltejs/sapper/issues/486) as well as this [issue](https://github.com/sveltejs/sapper/issues/278) . 

This is on a fresh setup as detailed [here](https://sapper.svelte.dev/docs#Getting_started). 

## Environment   

`Node Version: v8.16.0` , `OS: Arch Linux` , `npm version : 6.9.0` .

## Setting up (server.js) 

The default template comes with the following: 

```js 
import sirv from 'sirv';
import polka from 'polka';
import compression from 'compression';
import * as sapper from '@sapper/server';

const { PORT, NODE_ENV } = process.env;
const dev = NODE_ENV === 'development';

polka() // You can also use Express
	.use(
		compression({ threshold: 0 }),
		sirv('static', { dev }),
		sapper.middleware()
	)
	.listen(PORT, err => {
		if (err) console.log('error', err);
	}); 
```  

Then running `npm install` 

## Running 

`export HOST=127.0.0.1 PORT=3000`  

`npm run build` 

`npm run dev`  you get the following error 

```
 thebeachmaster@127  ~/test/svelte/hellosvelte  npm run dev                                                                                                                                  ✔  7070  10:50:04

> TODO@0.0.1 dev /home/thebeachmaster/test/svelte/hellosvelte
> sapper dev

✔ server (995ms)
✔ client (998ms)
✔ service worker (24ms)
> Server is not listening on port 3000
``` 
I went and checked if there was any process listening  on port 3000
`lsof -i :3000 | grep LISTEN` only to find that a `node` process was ... 

``` 
 thebeachmaster@thebeachmaster  ~/test/svelte/hellosvelte  lsof -i :3000 | grep LISTEN                                                                                                   1|1 ↵  7073  10:52:10
node    314 thebeachmaster   18u  IPv6 382263      0t0  TCP *:hbci (LISTEN)
 thebeachmaster@thebeachmaster  ~/test/svelte/hellosvelte   
```  

as seen here when I opened the browser...

![Sapper project template](https://user-images.githubusercontent.com/9129843/59156507-4b5af580-8aa5-11e9-8c1b-37719314c14e.png)

## The problem 
This results in hot-reloading failing.

Note that in my case as opposed to [this comment](https://github.com/sveltejs/sapper/issues/278#issuecomment-394499820) I did not have any other process on port 3000 other than the  app. 

## Attempts at solving

I went ahead and set `PORT=3080` and rebuilt the app, only to get the same error message but now on port `3080` ... but the app is running on port 3080 on localhost with port 3000 not listening on any processes (as expected).

``` thebeachmaster@127  ~/test/svelte/hellosvelte  npm run dev                                                                                                                                  ✔  7076  11:05:13

> TODO@0.0.1 dev /home/thebeachmaster/test/svelte/hellosvelte
> sapper dev

✔ server (959ms)
✔ client (962ms)
✔ service worker (22ms)
> Server is not listening on port 3080
```

**Comment by TheBeachMaster** [comment]
I have even switched to `express`  and still no change

```js
import compression from 'compression';
import * as sapper from '@sapper/server';
const express       = require('express');
const app = express();

app.use(express.static(__dirname + '/static'),
		compression({ threshold: 0 }),
		sapper.middleware()); 

const appPort = 3056;

app.listen(process.env.PORT || appPort, () => {
	console.info('We are up!');
	}); 
```

**Comment by Conduitry** [comment]
I've had this happen to me occasionally as well, generally if the app takes a long time to boot up. There's some timeout that stops waiting for the server to be ready, and displays this message. Increasing the timeout doesn't really seem like the proper solution.

I'm not sure how long we should wait for the server to be ready, but it does seem to make sense to start up the live reload server regardless. Or maybe we could just keep polling to see whether the server has started. There's definitely room for some improvement here.

**Comment by benoror** [comment]
Experiencing the same when using `.env` (dotenv) to set `PORT` env var to a different port number

**Comment by tbaums** [comment]
If anyone has a workaround, it would be much appreciated. Hitting the same issue.

**Comment by benoror** [comment]
I ended up starting the server while seting the `PORT` env var explicitly, as it seems `server.js` doesn't load `.env`:

```
$ PORT=3003 yarn run dev
```

Hope it helps!

**Comment by TheBeachMaster** [comment]
@benoror, Thanks for the suggestion. Unfortunately, this did not work for me. :(

**Comment by himynameisdave** [comment]
Question: why doesn't sapper just use something like [`get-port`](https://github.com/sindresorhus/get-port) to avoid collisions and ensure that it starts on an available port?

**Comment by thgh** [comment]
I have a collegue with this issue, the app is available but livereload does not work. Also the build takes longer than on my computer. It might have to do with storing the project in a network drive.

Still investigating.

**Comment by matt3224** [comment]
I've just run into this too not sure why, internet isn't too bad 60mbps. Seems to happen on all sapper projects too

**Comment by fratellomio** [comment]
I run into this issue too

**Comment by anshul2209** [comment]
Same issue here!

**Comment by mrmoree** [comment]
Any fix?

**Comment by TheBeachMaster** [comment]
There seems to be an open PR #985  

@Conduitry  any chance this will be looked at, seems like tons of people are having this issue 🤔

**Comment by znorris** [comment]
I run into this bug 100% of the time I try to use vscode's devcontainer to run a debug session.

**Comment by dilignt** [comment]
I am getting this very regularly.  Has anyone got a workaround?

**Comment by jdgaravito** [comment]
same as me, specially when sapper is trying many times to start the server upon any error

**Comment by marcusnewton** [comment]
Yeah I'm getting this. Seems to occur after stopping process with CTRL+C and running again. I guess we have to restart our PCs in the meantime?

**Comment by roblevintennis** [comment]
I have a use case to switch ports because  I have a work app that's on 3000 but I'd like to sometimes work on my blog on my breaks. It's a PITA because even though I've set up Svelte/Sapper app on 4000 something is opening it on 3000 too still:

![image](https://user-images.githubusercontent.com/142403/83420057-e3e73580-a3da-11ea-86fd-79e678b8b740.png)

The result of this is it's tedious to switch back to work project because I pretty much have to nuke the service worker even though I've stopped sapper server.

**Comment by elleeott** [comment]
I was having this issue all of a sudden, and it appears it was because I accidentally enabled two network interfaces simultaneously (wifi on and ethernet plugged in at same time). Disabling one or the other solved my problem. Not sure if that's helpful or not, but thought I'd share.

**Comment by BlakeMScurr** [comment]
I got this issue when I bumped `rollup-plugin-svelte` to `6.1.1` to solve a separate issue described here https://github.com/Samuel-Martineau/generator-svelte/issues/7. I'm also unsure if this is helpful or not.

**Comment by aslak01** [comment]
Having this issue trying to run Sapper on my Raspberry pi zero w. It's a bit weird, as I'm able to see the default output with curl localhost:3000, but  my Ngrok tunnel is giving an NSURLErrorDomain -1017, ie. not receiving valid data.

**Comment by Benjamin15122** [comment]
Got this issue with express@4.16.1. I figured out that it was because the usage of **app.listen()** in app.js and **"node /bin/www"** in package.json at the same time, since the default /bin/www would create a httpserver on 3000 anyway, even if app.listen has already created one. Remove app.listen() from app.js and change "node /bin/www" to "PORT=${port} node /bin/www" would fix the issue. Hope it helps :p.

**Comment by matheust3** [comment]
> I run into this bug 100% of the time I try to use vscode's devcontainer to run a debug session.


I also started to suffer from this problem after I configured the devcontainer in vscode

**Comment by luvnyen** [comment]
I was experiencing the same problem here. In my case, in the beginning, I was running the server with WSL on Windows 11. I have also tried changing the port several times and nothing worked. After that, I tried to run it without WSL and finally got this alert:
![messageImage_1643893001126](https://user-images.githubusercontent.com/56993480/152359958-ce3f6f12-389e-41cd-87f0-08b6795a7999.jpg)
The alert clearly states that "Window Defender Firewall has blocked some features of this app". In this case, "app" is the "Node.js: Server-side JavaScript" with the specific version of the currently used node version (in my case, it's v14.16.1).  So, I just click the "Allow access" button, and then it worked!

So, if you are using Windows, maybe you can try to [allow the connections in Windows Firewall manually](https://docs.profoundlogic.com/display/PUI/Allowing+Connections+in+Windows+Firewall). Hope it helps!
